### PR TITLE
Revocation support in policies

### LINKED ISSUE #102: Use annotations in RSL queries and policy verification
RSL annotations are distinct from regular entries in that they apply to other RSL entries. An annotation can also be marked as "skip" to indicate the referred RSL entries have been revoked. While it's possible to record RSL annotations, they're largely not used in the RSL package, nor in the policy package.

This is also a bug because unexpected annotations right now can lead to panics, crashing gittuf.

### PR DESCRIPTION
**Pull request by adityasaky**
~~Blocked by #149. This PR includes a commit from that PR.~~

Ready for review!

Closes #102

### CODE REVIEW

**Comment by adityasaky** [comment]
The revocation workflow needs to be better documented but roughly:

If an entry is determined to be invalid, the workflow checks that the entry has a skip annotation. Next, the workflow determines the last good entry for the ref. Then, the workflow scans through the rest of the entry queue to find an entry for the ref that is either commit-same or tree-same as the commit recorded in the last good entry for the ref. If one isn't found, verification fails. If this "fix" is found, then verification proceeds with the rest of the entry queue.

For the entries between the invalid entry and the fix entry, if they are not for the affected ref, they are left to be processed in future iterations. If the entry is for the affected ref, the workflow treats it as invalid and expects a skip annotation for it. If a skip annotation is not found, an error is returned. Note that in this part, the workflow essentially creates a new queue that takes the place of the original queue. The new queue is composed of entries of other refs between the invalid and fix entries as well as all the entries irrespective of the ref after the fix entry.

**Comment by adityasaky** [comment]
After #141 and #142 land, this PR will get smaller. Also, I'll submit the change to find unskipped RSL entries separately as well, with this one focused on the policy package revocation explicitly.

File: internal/policy/verify.go
Comment (adityasaky) [reply internal/policy/verify.go:171]: The full list of entries is provided by `rsl.GetReferenceEntriesInRangeForRef`. It returns ref entries for the requested ref and any entries for the gittuf namespace. I think preserving entries that aren't for the affected ref to be processed separately is fine in general because we may one day open this up to verifying multiple refs and not just one?

File: internal/policy/verify.go
Comment (adityasaky) [reply internal/policy/verify.go:147]: I can add a "trust but verify" check here!

File: internal/policy/verify.go
Comment (adityasaky) [reply internal/policy/verify.go:212]: That's fair. The idea is that we add a) all entries for other refs to the new queue, and b) we append contents of the current queue after a fix to this new queue. So the new queue is composed exclusively of entries that need to be verified still. This is because every reference entry for the affected ref between the first invalid entry and the fix entry are also considered to be invalid and must be skipped. So all of these are excluded from the new queue. What's a good way to feel more secure about this?

> especially given we're applying extra filtering in the lastGoodEntry check

Can you clarify this re the extra filtering? Do you mean the before clause?

File: internal/policy/verify.go
Comment (adityasaky) [reply internal/policy/verify.go:164]: I've switched this to more generally track invalid entries and error out if there are any. I think this is easier than the double negative boolean tracker, and also more extensible for a future where we verify multiple refs at once and we can track which ref continued to be unfixed etc.

File: internal/policy/verify.go
Comment (adityasaky) [reply internal/policy/verify.go:147]: Done!

File: internal/policy/verify.go
Comment (adityasaky) [reply internal/policy/verify.go:183]: I've also updated this conditional block to check that the candidate for the fix entry isn't also skipped. There's now also a test for this where the recovery passes, then we add an annotation skipping the recovery entry causing verification to fail again.

File: internal/policy/verify.go
Comment (wlynch) [reply internal/policy/verify.go:219]: Resolved off thread - we're appending all other entries about after confirming the sub-entry hasn't been skipped.

File: internal/policy/verify.go
Comment (wlynch) [reply internal/policy/verify.go:183]: Resolved off thread - this is used to match fix commits to previous ones using the tree state.

**Review by wlynch — APPROVED** [review]
LGTM, pending submission for some of the added docs.

**Comment by adityasaky** [comment]
Added some documentation in the invalid state bits, also opened #200.

**Comment by adityasaky** [comment]
#200 has been merged, @wlynch please merge if the doc updates here look good!

**Review by wlynch — APPROVED** [review]

**Pull request merged**
